File: examples/javascript/send.js
// Node.js >= 18 — uses the built-in fetch API. No npm install required.
//
// Usage:
//   DISCORD_WEBHOOK=https://discord.com/api/webhooks/.../...  node send.js

const WEBHOOK = process.env.DISCORD_WEBHOOK;
if (!WEBHOOK) {
  console.error('Set DISCORD_WEBHOOK env var first.');
  process.exit(1);
}

async function send(payload) {
  const res = await fetch(WEBHOOK, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(payload),
  });
  if (res.status === 429) {
    const body = await res.json();
    const waitMs = Math.ceil((body.retry_after || 1) * 1000) + 50;
    console.warn(`429 — sleeping ${waitMs}ms then retrying`);
    await new Promise((r) => setTimeout(r, waitMs));
    return send(payload);
  }
  if (!res.ok) {
    console.error(`HTTP ${res.status}`, await res.text());
  }
}

const embed = {
  title: 'Build failed',
  description: '`npm test` exited with code 1 on commit `a1b2c3d`.',
  color: 0xff0000,
  fields: [
    { name: 'Branch', value: 'main', inline: true },
    { name: 'CI', value: 'github-actions', inline: true },
    { name: 'Author', value: '@jay', inline: true },
  ],
  footer: { text: 'ci-bot' },
  timestamp: new Date().toISOString(),
};

await send({
  username: 'ci-bot',
  content: '<@&123456789012345678>', // role ping
  allowed_mentions: { parse: ['roles'] },
  embeds: [embed],
});

console.log('sent');

File: examples/rust/send.rs
// Rust — uses `reqwest` (async) + `tokio` + `serde_json`.
//
// Cargo.toml:
//   [dependencies]
//   reqwest = { version = "0.12", features = ["json"] }
//   tokio = { version = "1", features = ["full"] }
//   serde_json = "1"
//
// Usage:
//   DISCORD_WEBHOOK=https://discord.com/api/webhooks/.../...  cargo run

use serde_json::json;
use std::env;
use std::time::Duration;
use tokio::time::sleep;

async fn send_embed(
    client: &reqwest::Client,
    webhook: &str,
    payload: &serde_json::Value,
    max_retries: u32,
) -> Result<(), Box<dyn std::error::Error>> {
    for attempt in 1..=max_retries {
        let res = client.post(webhook).json(payload).send().await?;
        let status = res.status();
        if status.is_success() {
            return Ok(());
        }
        if status.as_u16() == 429 {
            let body: serde_json::Value = res.json().await.unwrap_or(json!({}));
            let wait = body
                .get("retry_after")
                .and_then(|v| v.as_f64())
                .unwrap_or(1.0);
            sleep(Duration::from_millis(((wait + 0.05) * 1000.0) as u64)).await;
            continue;
        }
        if status.is_server_error() && attempt < max_retries {
            sleep(Duration::from_secs(1)).await;
            continue;
        }
        let text = res.text().await.unwrap_or_default();
        return Err(format!("HTTP {}: {}", status, text).into());
    }
    Err("retries exhausted".into())
}

#[tokio::main]
async fn main() -> Result<(), Box<dyn std::error::Error>> {
    let webhook = env::var("DISCORD_WEBHOOK").expect("Set DISCORD_WEBHOOK");
    let client = reqwest::Client::builder()
        .timeout(Duration::from_secs(10))
        .build()?;

    let payload = json!({
        "username": "rust-bot",
        "embeds": [{
            "title": "Sensor reading critical",
            "description": "Tank #4 temperature 92°C (limit 85°C)",
            "color": 0xff0000,
            "fields": [
                {"name": "Sensor", "value": "tank-4-temp", "inline": true},
                {"name": "Threshold", "value": "85°C", "inline": true}
            ],
            "footer": {"text": "iot-monitor"},
            "timestamp": chrono_lite()
        }]
    });

    send_embed(&client, &webhook, &payload, 5).await?;
    println!("sent");
    Ok(())
}

fn chrono_lite() -> String {
    let secs = std::time::SystemTime::now()
        .duration_since(std::time::UNIX_EPOCH)
        .unwrap()
        .as_secs();
    // Crude ISO-8601 stringification (use `chrono` crate for production).
    format!("@{}", secs)
}
